File: src/Search/breadth-first_search.go
package Search

// O(V + E) - people + edges.
func BFS(graph map[string][]string, start string, condition func(string) bool) string {

	queue := []string{start}
	checked := []string{}

OUTER:
	for len(queue) > 0 {
		edge := queue[:1][0]
		queue = append(queue[:0], queue[0+1:]...)

		for _, v := range checked {
			if v == edge {
				continue OUTER
			}
		}
		checked = append(checked, edge)

		if condition(edge) {
			return edge
		}

		for _, v := range graph[edge] {
			queue = append(queue, v)
		}
	}
	return ""
}

File: src/Search/BreadthFirstSearch.php
<?php

namespace GulDmitry\Algorithms\Search;

/**
 * O(V + E) - people + edges.
 * Node: a block of information in the network.
 * Edge: a connection between two nodes (can have a direction and a weight).
 * Centrality: determining the relative importance of a node.
 * Clustering: partitioning nodes into groups.
 */
class BreadthFirstSearch
{
    public function search(array $graph, string $start, callable $condition): ?string
    {
        $queue = new \SplQueue();
        $queue->push($start);

        $checked = [];
        while (!$queue->isEmpty()) {
            $edge = $queue->shift();

            if (in_array($edge, $checked)) {
                continue;
            }
            $checked[] = $edge;

            if ($condition($edge)) {
                return $edge;
            }

            $rel = $graph[$edge];
            foreach ($rel as $node) {
                $queue->push($node);
            }
        }
        return null;
    }
}
